scheduler/docker/docker.go

```diff
@@ -1,6 +1,182 @@
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/fsouza/go-dockerclient"
+	"github.com/remind101/empire/pkg/dockerutil"
+	"github.com/remind101/empire/pkg/runner"
+	"github.com/remind101/empire/scheduler"
+	"golang.org/x/net/context"
+)
+
+type dockerClient interface {
+	InspectContainer(string) (*docker.Container, error)
+	ListContainers(docker.ListContainersOptions) ([]docker.APIContainers, error)
+}
+
+// Label that determines whether the container is from an attached run or not.
+const attachedRunLabel = "attached-run"
+
+// ListContainer filter to filter only attached runs.
+var attachedRunFilter = map[string][]string{
+	"label": []string{attachedRunLabel},
+}
+
+// attachedScheduler wraps a Scheduler to run attached processes using the Docker
+// scheduler.
+type attachedScheduler struct {
+	scheduler.Scheduler
+	dockerScheduler *Scheduler
+}
+
+// RunAttachedWithDocker wraps a Scheduler to run attached Run's using a Docker
+// client.
+func RunAttachedWithDocker(s scheduler.Scheduler, client *dockerutil.Client) scheduler.Scheduler {
+	return &attachedScheduler{
+		Scheduler:       s,
+		dockerScheduler: NewScheduler(client),
+	}
+}
+
+// Run runs attached processes using the docker scheduler, and detached
+// processes using the wrapped scheduler.
+func (s *attachedScheduler) Run(ctx context.Context, app *scheduler.App, process *scheduler.Process, in io.Reader, out io.Writer) error {
+	// Attached means stdout, stdin is attached.
+	attached := out != nil || in != nil
+
+	if attached {
+		return s.dockerScheduler.Run(ctx, app, process, in, out)
+	} else {
+		return s.Scheduler.Run(ctx, app, process, in, out)
+	}
+}
+
+// Instances returns a combination of instances from the wrapped scheduler, as
+// well as instances from attached runs.
+func (s *attachedScheduler) Instances(ctx context.Context, app string) ([]*scheduler.Instance, error) {
+	instances, err := s.Scheduler.Instances(ctx, app)
+	if err != nil {
+		return instances, err
+	}
+
+	attachedInstances, err := s.dockerScheduler.InstancesFromAttachedRuns(ctx, app)
+	if err != nil {
+		return instances, err
+	}
+
+	return append(instances, attachedInstances...), nil
+}
+
+// Scheduler provides an implementation of the scheduler.Scheduler interface
+// backed by Docker.
+type Scheduler struct {
+	runner *runner.Runner
+	docker dockerClient
+}
+
+// NewScheduler returns a new Scheduler instance that uses the given client to
+// interact with Docker.
+func NewScheduler(client *dockerutil.Client) *Scheduler {
+	return &Scheduler{
+		runner: runner.NewRunner(client),
+		docker: client,
+	}
+}
+
+func (s *Scheduler) Run(ctx context.Context, app *scheduler.App, p *scheduler.Process, in io.Reader, out io.Writer) error {
+	attached := out != nil || in != nil
+
+	if !attached {
+		return errors.New("cannot run detached processes with Docker scheduler")
+	}
+
+	labels := scheduler.Labels(app, p)
+	labels[attachedRunLabel] = p.Type
+	return s.runner.Run(ctx, runner.RunOpts{
```

Comment:
Might as well just pull `pkg/runner` in here, since it's only used inside `scheduler/docker`.